### DISCUSSION TITLE
getStaticPaths fallback: 'blocking' causes page to break because props are empty

### DISCUSSION THREAD

**Discussion by MartinaMilisic** [vercel/next.js#47772] [Help] [answered]
### Summary

Hi!

I'm on Next.js 13.1.6 and I'm confused on how the `fallback: blocking` option in `getStaticProps` should work. Everything runs great if I build the project and try to run it locally, but when deploying it to azure web apps, I get an error on the pages that are not pre generated with `getStaticPaths`, that the props are empty. 

As I understand it, fallback blocking will wait until getStaticProps has run, and then send the data to the page. I've added the `router.isFallback` check provided in the documentation, but it seems the fallback is false. Also, I understand it as this check should not be needed with fallback: blocking, only with fallback: true.

I'm using a catch all route to fetch data from Sanity cms:

```
const Slug: NextPageWithLayout<IPage> = ({ pageData }) => {
	const router = useRouter()
	console.log('fallback', router.isFallback)

	// If the page is not yet generated, this will be displayed
	// initially until getStaticProps() finishes running
	if (router.isFallback) {
		return <div>Loading...</div>
	}

	if (!pageData) {
		// This is what I get
		return <h1>No props</h1>
	}

	const { title, _type } = pageData
	const pageTitle = title ?? 'Page'
	const pageTemplate = getContentByType(_type)

	return (
		<>
			<Head>
				<title>{pageTitle}</title>
			</Head>
			{/* Render page by documentType */}
			{_type && React.createElement(pageTemplate, pageData)}
		</>
	)
}

export const getStaticPaths: GetStaticPaths = async ({ locales }) => {
	//TODO:  generate slugs
	return { paths: [], fallback: 'blocking' }
}

export const getStaticProps: GetStaticProps = async ({ params }) => {
	const { slug } = params!

	const joinedSlug = typeof slug === 'string' ? slug : slug?.join('/')
	const data: IGetPageData = await getPageData(`/${joinedSlug!}`)

	return {
		props: {
			megaMenuData: headerData,
			footerData: data.footerData,
			pageData: { ...data.pageData, breadcrumbs: { ...data.breadcrumbs } }
		},
		revalidate: 60
	}
}

export default Slug
```

I'm using the experimental feature in my next.config.js to be able to  get the right output when using the standalone feature on build in my monorepo:

```
output: 'standalone',
experimental: {
outputFileTracingRoot: path.join(__dirname, '../../')
}

```
Any help would be much appreciated! Thank you :)

**Comment by icyJoseph** [comment]
Is `getPageData` working correctly in your deployment? What does that function do?

**Reply by MartinaMilisic** [reply]
That was my initial thought as well, but if i reload the page, the page will display properly, just as expected with blocking fallback (I assume). Also, everything works when building the page locally running `next build` and `next start`. 

The getPageData looks like this:

```
export const getPageData = async (slug: string) => {
	const data = await getInitialPageData(slug)

	return data
}
```

And the getInitialPageData looks like this:
```
export const getInitialPageData = async (slug: string) => {
	let data
	try {
		data = await client.fetch(
			`{
			${metaMenuData},
			${megaMenuData},
			${footerData},
			"breadcrumbs": *[slug.current == $slug || editSlug.slug.current == $slug ] | order(_updatedAt desc) {
				title,
				parent->{
					title,
					"url": select(defined(slug.current) => slug.current, defined(editSlug.slug.current) => editSlug.slug.current),
					parent->{
						title,
						"url": select(defined(slug.current) => slug.current, defined(editSlug.slug.current) => editSlug.slug.current),
						parent->{
							title,
							'url': select(defined(slug.current) => slug.current, defined(editSlug.slug.current) => editSlug.slug.current),
						}
					}
				}
			}[0],
			"pageData": *[slug.current == $slug || editSlug.slug.current == $slug ] | order(_updatedAt desc) {
					_type,
					title,
					description,
					seo
				}[0]
			}`,
			{ slug }
		)
	} catch (error) {
		console.warn(error)
	}

	return data
}
```

**Reply by icyJoseph** [reply]
Right, and you see no logs in the Azure environment? My guess would be that there's some keys, or env vars missing, and this `getInitialPageData` function is kind of letting failure resolve to no-data (`undefined`), which seems to match your issue. 

Might also be worth making sure that the Node.js version is the same in the deployment and in your local machine.

**Reply by MartinaMilisic** [reply]
Thanks for the thought! Yeah, I think (hope) that might be the case as well. I didn't have access to the logs when writing the question. Hopefully I can investigate more later this week. I'll get back when I know more. Thank you again!

**Reply by icyJoseph** [reply]
Great! Another tip would be to return some additional data (like, error: "something went wrong") in the props for the catch scenario, that way we know the data query is the faulty part, and not something with Next.js -

**Reply by MartinaMilisic** [reply]
Great tip! You mean something like this?

<img width="615" alt="image" src="https://user-images.githubusercontent.com/125043449/229485092-3cc7af02-bc96-4866-99c2-82fd8c69eb6e.png">

<img width="650" alt="image" src="https://user-images.githubusercontent.com/125043449/229485143-a9cc5369-5c04-4223-a72a-696bda4b523a.png">

**Reply by icyJoseph** [reply]
Almost! You can verify if error is `instanceof` Error, and then take the message, otherwise just try to return it as you have done.

**Reply by MartinaMilisic** [reply]
You mean in the try catch or in getStaticProps?

**Reply by icyJoseph** [reply]
In the try catch, and then continue do as you did, in the getStaticProps

**Reply by MartinaMilisic** [reply]
Thanks!

**Comment by roger-swipejobs** [comment]
Are you running in a Next.js server ou in a s3 bucket? I think the next server must be running, maybe the output: standalone generate files as static to run in a s3 bucket. for example

**Reply by icyJoseph** [reply]
standalone is a server, it outputs a server.js file that you run - so it might just be that, perhaps they meant to do `export` and not `standalone` - good observations

**Comment by MartinaMilisic** [comment, marked as answer]
@icyJoseph and @roger-swipejobs Thanks for your answers! I forgot about this thread :) We figured out in the end what the problem seems to be. The pages that generated an error, where pages that we previously had deployed to azure, but that was not entirely working (not getting the right data or what not). Azure seems to be caching all files and hence it returned an old file the first time. I thought this might be the next cache, but when emptying the artifacts pushed to azure before the next deploy, the right files where showing from start. Maybe this is as intended, but it seemed a bit annoying to me :) Anyway, thank you for your thoughts!
